File: app/src/main/java/com/qadam/lastpuff/ui/screens/sos/SosScreen.kt
package com.qadam.lastpuff.ui.screens.sos

import androidx.compose.foundation.layout.Row
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FilterChip
import androidx.compose.material3.FilterChipDefaults
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.qadam.lastpuff.ui.screens.body.BodyVictoryScreen
import com.qadam.lastpuff.ui.components.BreathingCircle
import com.qadam.lastpuff.ui.components.SosFullscreenBackground
import com.qadam.lastpuff.ui.viewmodel.AppViewModel
import com.qadam.lastpuff.ui.viewmodel.SosStep
import com.qadam.lastpuff.util.AppConstants
import com.qadam.lastpuff.util.ContactUtils
import kotlinx.coroutines.delay
import java.util.Locale

@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@Composable
fun SosScreen(
    viewModel: AppViewModel,
    onBack: () -> Unit,
    onComplete: () -> Unit
) {
    val state by viewModel.sosState.collectAsState()
    val sosContact by viewModel.sosContact.collectAsState()
    val familyPhotoUri by viewModel.familyPhotoUri.collectAsState()
    val personalLetter by viewModel.personalLetter.collectAsState()
    val homeStats by viewModel.homeStats.collectAsState()
    val context = LocalContext.current

    LaunchedEffect(state.step) {
        if (state.step == SosStep.TIMER) {
            while (true) {
                delay(1000)
                if (viewModel.sosState.value.step != SosStep.TIMER) break
                viewModel.tickTimer()
            }
        }
    }

    LaunchedEffect(state.step) {
        if (state.step == SosStep.BREATHING) {
            while (true) {
                delay(4000)
                if (viewModel.sosState.value.step != SosStep.BREATHING) break
                viewModel.tickBreathing()
            }
        }
    }

    LaunchedEffect(state.step, state.showCoinAnimation) {
        if (state.step == SosStep.RELAPSE) {
            delay(3500)
            viewModel.resetSos()
            onComplete()
        }
    }

    when (state.step) {
        SosStep.EMERGENCY -> EmergencyScreen(
            contact = sosContact,
            onCall = { sosContact?.let { ContactUtils.call(context, it.phone) } },
            onSms = { sosContact?.let { ContactUtils.sms(context, it.phone, it.message) } },
            onWhatsApp = { sosContact?.let { ContactUtils.whatsApp(context, it.phone, it.message) } },
            onTelegram = { sosContact?.let { ContactUtils.telegram(context, it.phone, it.message) } },
            onBreathing = { viewModel.startBreathing() },
            onTimer = { viewModel.startTimerFromEmergency() },
            onBack = { viewModel.resetSos(); onBack() }
        )

        SosStep.TIMER, SosStep.BREATHING -> SosFullscreenBackground {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(24.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Row(modifier = Modifier.fillMaxWidth()) {
                    IconButton(
                        onClick = {
                            if (state.step == SosStep.BREATHING) {
                                viewModel.resumeTimer()
                            }
                        },
                        enabled = state.step == SosStep.BREATHING
                    ) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = "К таймеру",
                            tint = Color.White.copy(alpha = if (state.step == SosStep.BREATHING) 1f else 0f)
                        )
                    }
                }

                if (state.step == SosStep.BREATHING) {
                    Spacer(modifier = Modifier.weight(1f))
                    BreathingCircle(phase = state.breathingPhase)
                    Spacer(modifier = Modifier.weight(1f))
                } else {
                    Spacer(modifier = Modifier.weight(0.12f))

                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(210.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.Center
                    ) {
                        val minutes = state.timerSecondsLeft / 60
                        val seconds = state.timerSecondsLeft % 60
                        Text(
                            text = "НЕ ЗАКРЫВАЙ",
                            style = MaterialTheme.typography.labelLarge,
                            color = Color.White.copy(alpha = 0.8f),
                            letterSpacing = 2.sp
                        )
                        Text(
                            text = "Я С ТОБОЙ",
                            style = MaterialTheme.typography.titleMedium,
                            color = Color.White.copy(alpha = 0.9f)
                        )
                        Spacer(modifier = Modifier.height(16.dp))
                        Text(
                            text = String.format(Locale.getDefault(), "%d:%02d", minutes, seconds),
                            style = MaterialTheme.typography.displayLarge.copy(
                                fontSize = 72.sp,
                                fontWeight = FontWeight.Bold,
                                color = Color.White
                            )
                        )
                        Spacer(modifier = Modifier.height(12.dp))
                        LinearProgressIndicator(
                            progress = { state.elapsedSeconds / AppConstants.SOS_TIMER_SECONDS.toFloat() },
                            modifier = Modifier
                                .fillMaxWidth()
                                .height(6.dp)
                                .clip(RoundedCornerShape(3.dp)),
                            color = Color.White,
                            trackColor = Color.White.copy(alpha = 0.3f)
                        )
                    }

                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(210.dp)
                            .padding(horizontal = 4.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.Top
                    ) {
                        Box(
                            modifier = Modifier
                                .fillMaxWidth()
                                .height(96.dp),
                            contentAlignment = Alignment.Center
                        ) {
                            Text(
                                text = state.currentMessage,
                                style = MaterialTheme.typography.titleMedium,
                                color = Color.White,
                                textAlign = TextAlign.Center,
                                lineHeight = 26.sp,
                                maxLines = 3
                            )
                        }
                        Box(
                            modifier = Modifier
                                .fillMaxWidth()
                                .height(72.dp),
                            contentAlignment = Alignment.Center
                        ) {
                            state.secondaryMessage?.let { secondary ->
                                Text(
                                    text = secondary,
                                    style = MaterialTheme.typography.bodyLarge,
                                    color = Color.White.copy(alpha = 0.88f),
                                    textAlign = TextAlign.Center,
                                    lineHeight = 24.sp,
                                    maxLines = 2
                                )
                            }
                        }
                        Box(
                            modifier = Modifier
                                .fillMaxWidth()
                                .height(40.dp),
                            contentAlignment = Alignment.Center
                        ) {
                            state.currentAction?.let { action ->
                                Text(
                                    text = when (state.sosMode) {
                                        com.qadam.lastpuff.domain.support.SosMode.BREATHING -> "🌬 $action"
                                        com.qadam.lastpuff.domain.support.SosMode.FAMILY -> "❤️ $action"
                                        com.qadam.lastpuff.domain.support.SosMode.CHALLENGE -> "🎯 $action"
                                    },
                                    style = MaterialTheme.typography.bodyMedium,
                                    color = Color.White.copy(alpha = 0.85f),
                                    textAlign = TextAlign.Center,
                                    maxLines = 1
                                )
                            }
                        }
                    }

                    FlowRow(
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        SosActionChip("Дыхание") { viewModel.startBreathing() }
                        SosActionChip("Причины") { viewModel.showReasons() }
                        if (familyPhotoUri != null) {
                            SosActionChip("Фото") { viewModel.showPhoto() }
                        }
                        sosContact?.let {
                            SosActionChip("Позвонить") { ContactUtils.call(context, it.phone) }
                            SosActionChip("WhatsApp") { ContactUtils.whatsApp(context, it.phone, it.message) }
                        }
                    }
                    Spacer(modifier = Modifier.height(12.dp))
                    TextButton(onClick = { viewModel.showOptionalDetails() }) {
                        Text("Если хочешь — укажи причину", color = Color.White.copy(alpha = 0.7f))
                    }
                    Spacer(modifier = Modifier.weight(0.08f))
                }
            }
        }

        SosStep.RESULT -> SosFullscreenBackground {
            Column(
                modifier = Modifier.padding(32.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center
            ) {
                Text(
                    text = "Удалось не закурить?",
                    style = MaterialTheme.typography.headlineMedium,
                    color = Color.White,
                    textAlign = TextAlign.Center
                )
                Spacer(modifier = Modifier.height(32.dp))
                Button(
                    onClick = { viewModel.completeSos(true) },
                    modifier = Modifier.fillMaxWidth(),
                    colors = ButtonDefaults.buttonColors(containerColor = Color.White),
                ) {
                    Text("Да, справился", color = Color(0xFFB71C1C))
                }
                Spacer(modifier = Modifier.height(12.dp))
                OutlinedButton(
                    onClick = { viewModel.completeSos(false) },
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Text("Продолжаю путь", color = Color.White)
                }
            }
        }

        SosStep.SUCCESS -> BodyVictoryScreen(
            before = state.recoveryBefore,
            after = state.recoveryAfter,
            showCoinAnimation = state.showCoinAnimation,
            totalCoins = homeStats?.totalCoins ?: 1,
            coinsEarned = state.coinsEarned,
            onContinue = {
                viewModel.resetSos()
                onComplete()
            }
        )

        SosStep.RELAPSE -> SosFullscreenBackground {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                modifier = Modifier.padding(32.dp)
            ) {
                Text("💚", fontSize = 56.sp)
                Text(
                    text = state.currentMessage.ifBlank { "Ты продолжаешь путь" },
                    style = MaterialTheme.typography.headlineMedium,
                    color = Color.White,
                    textAlign = TextAlign.Center,
                    fontWeight = FontWeight.Bold
                )
                Spacer(modifier = Modifier.height(12.dp))
                Text(
                    text = "Один момент не отменяет\nвесь твой прогресс.",
                    style = MaterialTheme.typography.bodyLarge,
                    color = Color.White.copy(alpha = 0.9f),
                    textAlign = TextAlign.Center
                )
            }
        }
    }

    if (state.showOptionalDetails) {
        ModalBottomSheet(
            onDismissRequest = { viewModel.dismissOptionalDetails() },
            sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
        ) {
            Column(modifier = Modifier.padding(24.dp)) {
                Text("Если хочешь — укажи причину", style = MaterialTheme.typography.titleLarge)
                Spacer(modifier = Modifier.height(12.dp))
                FlowRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    AppConstants.CRAVING_TRIGGERS.forEach { trigger ->
                        FilterChip(
                            selected = state.trigger == trigger,
                            onClick = { viewModel.setOptionalTrigger(trigger) },
                            label = { Text(trigger) }
                        )
                    }
                }
                Spacer(modifier = Modifier.height(16.dp))
                Text("Сила тяги: ${state.intensity}")
                Slider(
                    value = state.intensity.toFloat(),
                    onValueChange = { viewModel.setOptionalIntensity(it.toInt()) },
                    valueRange = 1f..10f,
                    steps = 8
                )
                Spacer(modifier = Modifier.height(16.dp))
                Button(
                    onClick = { viewModel.dismissOptionalDetails() },
                    modifier = Modifier.fillMaxWidth()
                ) { Text("Готово") }
                Spacer(modifier = Modifier.height(24.dp))
            }
        }
    }

    if (state.showReasons) {
        val reasons = viewModel.quitReasons()
        AlertDialog(
            onDismissRequest = { viewModel.dismissReasons() },
            title = { Text("Почему нельзя сейчас") },
            text = {
                Column {
                    reasons.forEach { Text("• $it", modifier = Modifier.padding(vertical = 4.dp)) }
                }
            },
            confirmButton = {
                Button(onClick = { viewModel.dismissReasons() }) { Text("Продолжить") }
            }
        )
    }

    if (state.showPhoto && familyPhotoUri != null) {
        AlertDialog(
            onDismissRequest = { viewModel.dismissPhoto() },
            title = { Text("Ради них") },
            text = {
                AsyncImage(
                    model = familyPhotoUri,
                    contentDescription = "Фото близких",
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(240.dp)
                        .clip(RoundedCornerShape(16.dp)),
                    contentScale = ContentScale.Crop
                )
            },
            confirmButton = {
                Button(onClick = { viewModel.dismissPhoto() }) { Text("Продолжить") }
            }
        )
    }

    if (state.showPersonalLetterDialog && !personalLetter.isNullOrBlank()) {
        AlertDialog(
            onDismissRequest = { viewModel.dismissPersonalLetter() },
            title = { Text("Письмо себе") },
            text = {
                Text(
                    text = personalLetter!!,
                    style = MaterialTheme.typography.bodyLarge,
                    lineHeight = 24.sp
                )
            },
            confirmButton = {
                Button(onClick = { viewModel.dismissPersonalLetter() }) { Text("Продолжать путь") }
            }
        )
    }
}

@Composable
private fun EmergencyScreen(
    contact: com.qadam.lastpuff.domain.model.SosContact?,
    onCall: () -> Unit,
    onSms: () -> Unit,
    onWhatsApp: () -> Unit,
    onTelegram: () -> Unit,
    onBreathing: () -> Unit,
    onTimer: () -> Unit,
    onBack: () -> Unit
) {
    SosFullscreenBackground {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Text("🚨", fontSize = 48.sp)
            Text(
                text = "КРИТИЧЕСКАЯ\nТЯГА",
                style = MaterialTheme.typography.displaySmall,
                color = Color.White,
                textAlign = TextAlign.Center,
                fontWeight = FontWeight.Bold
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = "НЕ СДАВАЙСЯ",
                style = MaterialTheme.typography.headlineMedium,
                color = Color.White.copy(alpha = 0.9f)
            )
            Spacer(modifier = Modifier.height(32.dp))
            EmergencyButton("📞 Позвонить ${contact?.name ?: ""}", onCall)
            Spacer(modifier = Modifier.height(8.dp))
            EmergencyButton("💬 SMS", onSms)
            Spacer(modifier = Modifier.height(8.dp))
            EmergencyButton("WhatsApp", onWhatsApp)
            Spacer(modifier = Modifier.height(8.dp))
            EmergencyButton("Telegram", onTelegram)
            Spacer(modifier = Modifier.height(8.dp))
            EmergencyButton("🌬 Дыхание", onBreathing)
            Spacer(modifier = Modifier.height(8.dp))
            EmergencyButton("⏱ 3 минуты вместе", onTimer)
            Spacer(modifier = Modifier.height(16.dp))
            TextButton(onClick = onBack) {
                Text("Назад", color = Color.White.copy(alpha = 0.7f))
            }
        }
    }
}

@Composable
private fun EmergencyButton(text: String, onClick: () -> Unit) {
    Button(
        onClick = onClick,
        modifier = Modifier.fillMaxWidth(),
        colors = ButtonDefaults.buttonColors(containerColor = Color.White),
        shape = RoundedCornerShape(16.dp)
    ) {
        Text(text, color = Color(0xFFB71C1C), modifier = Modifier.padding(vertical = 4.dp))
    }
}

@Composable
private fun SosActionChip(label: String, onClick: () -> Unit) {
    FilterChip(
        selected = false,
        onClick = onClick,
        label = { Text(label, color = Color.White) },
        colors = FilterChipDefaults.filterChipColors(
            containerColor = Color.White.copy(alpha = 0.15f),
            labelColor = Color.White
        )
    )
}

File: ios/QadamLastPuff/Views/Screens/SosScreen.swift
import SwiftUI

struct SosScreen: View {
    @ObservedObject var viewModel: AppViewModel
    let onBack: () -> Void
    let onComplete: () -> Void

    @State private var timerTask: Task<Void, Never>?
    @State private var breathingTask: Task<Void, Never>?

    var body: some View {
        Group {
            switch viewModel.sosState.step {
            case .emergency:
                EmergencyScreen(
                    contact: viewModel.sosContact,
                    onCall: { if let c = viewModel.sosContact { ContactUtils.call(phone: c.phone) } },
                    onSms: { if let c = viewModel.sosContact { ContactUtils.sms(phone: c.phone, message: c.message) } },
                    onWhatsApp: { if let c = viewModel.sosContact { ContactUtils.whatsApp(phone: c.phone, message: c.message) } },
                    onTelegram: { if let c = viewModel.sosContact { ContactUtils.telegram(phone: c.phone, message: c.message) } },
                    onBreathing: { viewModel.startBreathing() },
                    onTimer: { viewModel.startTimerFromEmergency() },
                    onBack: { viewModel.resetSos(); onBack() }
                )
            case .intake:
                intakeView
            case .timer, .breathing:
                timerBreathingView
            case .result:
                resultView
            case .success:
                BodyVictoryScreen(
                    before: viewModel.sosState.recoveryBefore,
                    after: viewModel.sosState.recoveryAfter,
                    showCoinAnimation: viewModel.sosState.showCoinAnimation,
                    totalCoins: viewModel.homeStats?.totalCoins ?? 0,
                    onContinue: { viewModel.resetSos(); onComplete() }
                )
            case .relapse:
                relapseView
            }
        }
        .onAppear { startTimers() }
        .onDisappear { stopTimers() }
        .onChange(of: viewModel.sosState.step) { _, _ in startTimers() }
        .alert("Письмо себе", isPresented: Binding(
            get: { viewModel.sosState.showPersonalLetterDialog },
            set: { if !$0 { viewModel.dismissPersonalLetter() } }
        )) {
            Button("Понятно") { viewModel.dismissPersonalLetter() }
        } message: {
            Text(viewModel.personalLetter ?? "")
        }
        .alert("Почему ты бросаешь?", isPresented: Binding(
            get: { viewModel.sosState.showReasons },
            set: { if !$0 { viewModel.dismissReasons() } }
        )) {
            Button("Закрыть") { viewModel.dismissReasons() }
        } message: {
            Text(viewModel.quitReasons().joined(separator: "\n"))
        }
        .confirmationDialog("Выйти из таймера?", isPresented: Binding(
            get: { viewModel.sosState.showExitDialog },
            set: { if !$0 { viewModel.dismissSosExitDialog() } }
        ), titleVisibility: .visible) {
            Button("Таймер не нужен — выйти") {
                viewModel.exitSosWithoutTimer()
                onBack()
            }
            Button("Обнулить и начать заново") {
                viewModel.restartSosTimer()
            }
            Button("Продолжить таймер", role: .cancel) {
                viewModel.dismissSosExitDialog()
            }
        } message: {
            Text("Ты уже \(viewModel.sosState.elapsedSeconds) сек держишься. Что хочешь сделать?")
        }
        .sheet(isPresented: Binding(
            get: { viewModel.sosState.showPhoto },
            set: { if !$0 { viewModel.dismissPhoto() } }
        )) {
            VStack(spacing: 16) {
                Text("Ради них").font(.title2.bold())
                if let uri = viewModel.familyPhotoUri, let url = URL(string: uri) {
                    AsyncImage(url: url) { $0.resizable().scaledToFit() } placeholder: { ProgressView() }
                }
                Button("Закрыть") { viewModel.dismissPhoto() }
            }
            .padding()
        }
    }

    private var intakeView: some View {
        SosFullscreenBackground {
            ScrollView {
                VStack(alignment: .leading, spacing: 20) {
                    HStack {
                        Button {
                            viewModel.resetSos()
                            onBack()
                        } label: {
                            Image(systemName: "xmark")
                                .foregroundStyle(.white)
                                .padding(8)
                                .background(.white.opacity(0.2))
                                .clipShape(Circle())
                        }
                        Spacer()
                    }

                    Text("Расскажи, что происходит")
                        .font(.title2.bold())
                        .foregroundStyle(.white)
                    Text("Это поможет понять твои триггеры и стать сильнее")
                        .foregroundStyle(.white.opacity(0.8))

                    Text("Из-за чего?").font(.headline).foregroundStyle(.white)
                    ChipFlowView(
                        items: AppConstants.cravingTriggers,
                        selection: viewModel.sosState.trigger,
                        lightOnDark: true,
                        onSelect: { viewModel.setSosTrigger($0) }
                    )

                    Text("Где ты сейчас?").font(.headline).foregroundStyle(.white)
                    ChipFlowView(
                        items: AppConstants.cravingLocations,
                        selection: viewModel.sosState.location,
                        lightOnDark: true,
                        onSelect: { viewModel.setSosLocation($0) }
                    )

                    Text("Насколько сильная тяга: \(viewModel.sosState.intensity)")
                        .font(.headline)
                        .foregroundStyle(.white)
                    Slider(value: Binding(
                        get: { Double(viewModel.sosState.intensity) },
                        set: { viewModel.setSosIntensity(Int($0)) }
                    ), in: 1...10, step: 1)
                    .tint(.white)

                    Button("Начать 3 минуты вместе") {
                        viewModel.confirmSosIntakeAndStartTimer()
                    }
                    .buttonStyle(.borderedProminent)
                    .tint(.white)
                    .foregroundStyle(QadamColors.sosRedDark)
                    .frame(maxWidth: .infinity)
                    .disabled(viewModel.sosState.trigger.isEmpty)
                }
                .padding(24)
            }
        }
    }

    private var timerBreathingView: some View {
        SosFullscreenBackground {
            VStack(spacing: 0) {
                HStack {
                    Button {
                        if viewModel.sosState.step == .breathing {
                            viewModel.resumeTimer()
                        } else {
                            viewModel.showSosExitDialog()
                        }
                    } label: {
                        Image(systemName: "chevron.left")
                            .foregroundStyle(.white)
                            .padding(10)
                            .background(.white.opacity(0.2))
                            .clipShape(Circle())
                    }
                    Spacer()
                }
                .padding(.horizontal, 20)
                .padding(.top, 8)

                if viewModel.sosState.step == .breathing {
                    Spacer()
                    BreathingCircle(phase: viewModel.sosState.breathingPhase)
                    Spacer()
                } else {
                    timerHeaderSection
                        .padding(.horizontal, 20)
                        .padding(.top, 8)

                    timerMessageSection
                        .frame(height: 200)
                        .padding(.horizontal, 20)

                    Spacer(minLength: 12)

                    timerActionsSection
                        .padding(.horizontal, 20)
                        .padding(.bottom, 28)
                }
            }
        }
    }

    private var timerHeaderSection: some View {
        VStack(spacing: 8) {
            Text("НЕ ЗАКРЫВАЙ")
                .font(.caption)
                .tracking(2)
                .foregroundStyle(.white.opacity(0.8))
            Text("Я С ТОБОЙ")
                .foregroundStyle(.white.opacity(0.9))
            Text(timerText)
                .font(.system(size: 64, weight: .bold))
                .foregroundStyle(.white)
                .monospacedDigit()
            ProgressView(
                value: Double(viewModel.sosState.elapsedSeconds),
                total: Double(AppConstants.sosTimerSeconds)
            )
            .tint(.white)
        }
        .frame(maxWidth: .infinity)
    }

    private var timerMessageSection: some View {
        VStack(spacing: 0) {
            Text(viewModel.sosState.currentMessage)
                .font(.headline)
                .foregroundStyle(.white)
                .multilineTextAlignment(.center)
                .lineLimit(3)
                .minimumScaleFactor(0.85)
                .frame(maxWidth: .infinity)
                .frame(height: 96)

            Group {
                if let secondary = viewModel.sosState.secondaryMessage {
                    Text(secondary)
                        .foregroundStyle(.white.opacity(0.85))
                        .multilineTextAlignment(.center)
                        .lineLimit(2)
                        .minimumScaleFactor(0.85)
                } else {
                    Color.clear
                }
            }
            .frame(maxWidth: .infinity)
            .frame(height: 72)

            Group {
                if let action = viewModel.sosState.currentAction {
                    Text("\(viewModel.sosState.sosMode.emoji) \(action)")
                        .font(.subheadline)
                        .foregroundStyle(.white.opacity(0.9))
                        .lineLimit(1)
                } else {
                    Color.clear
                }
            }
            .frame(maxWidth: .infinity)
            .frame(height: 32)
        }
    }

    private var timerActionsSection: some View {
        let contact = viewModel.sosContact
        let hasPhone = PhoneMask.isComplete(contact?.phone ?? "")
        let message = contact?.message ?? ""

        return VStack(spacing: 10) {
            HStack(spacing: 10) {
                actionRow("Дыхание", compact: true) { viewModel.startBreathing() }
                actionRow("Причины", compact: true) { viewModel.showReasons() }
                if viewModel.familyPhotoUri != nil {
                    actionRow("Фото", compact: true) { viewModel.showPhoto() }
                } else {
                    actionPlaceholder(compact: true)
                }
            }

            HStack(spacing: 10) {
                actionRow("📞 Звонок", enabled: hasPhone, compact: true) {
                    if let phone = contact?.phone { ContactUtils.call(phone: phone) }
                }
                actionRow("💬 WhatsApp", enabled: hasPhone, compact: true) {
                    if let phone = contact?.phone { ContactUtils.whatsApp(phone: phone, message: message) }
                }
                actionRow("✈️ Telegram", enabled: hasPhone, compact: true) {
                    if let phone = contact?.phone { ContactUtils.telegram(phone: phone, message: message) }
                }
            }
        }
    }

    private func actionRow(
        _ title: String,
        enabled: Bool = true,
        compact: Bool = false,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            Text(title)
                .font(compact ? .subheadline.weight(.medium) : .body.weight(.medium))
                .foregroundStyle(.white.opacity(enabled ? 1 : 0.45))
                .lineLimit(1)
                .minimumScaleFactor(0.75)
                .frame(maxWidth: .infinity)
                .padding(.vertical, compact ? 10 : 12)
                .background(Color.white.opacity(enabled ? 0.2 : 0.08))
                .clipShape(RoundedRectangle(cornerRadius: 12))
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
    }

    private func actionPlaceholder(compact: Bool) -> some View {
        RoundedRectangle(cornerRadius: 12)
            .fill(Color.clear)
            .frame(maxWidth: .infinity)
            .frame(height: compact ? 38 : 44)
    }

    private var resultView: some View {
        SosFullscreenBackground {
            VStack(spacing: 24) {
                Text("Удалось не закурить?")
                    .font(.title2.bold())
                    .foregroundStyle(.white)
                Button("Да, справился") { viewModel.completeSos(success: true) }
                    .buttonStyle(.borderedProminent)
                    .tint(.white)
                    .foregroundStyle(QadamColors.sosRedDark)
                Button("Продолжаю путь") { viewModel.completeSos(success: false) }
                    .buttonStyle(.bordered)
                    .tint(.white)
            }
            .padding(24)
        }
    }

    private var relapseView: some View {
        SosFullscreenBackground {
            VStack(spacing: 16) {
                Text("💚").font(.largeTitle)
                Text(viewModel.sosState.currentMessage)
                    .font(.title3)
                    .foregroundStyle(.white)
                    .multilineTextAlignment(.center)
                Text("Один момент не отменяет весь твой прогресс.")
                    .foregroundStyle(.white.opacity(0.8))
            }
            .padding(24)
            .onAppear {
                DispatchQueue.main.asyncAfter(deadline: .now() + 3.5) {
                    viewModel.resetSos()
                    onComplete()
                }
            }
        }
    }

    private var timerText: String {
        let m = viewModel.sosState.timerSecondsLeft / 60
        let s = viewModel.sosState.timerSecondsLeft % 60
        return String(format: "%d:%02d", m, s)
    }

    private func startTimers() {
        stopTimers()
        if viewModel.sosState.step == .timer {
            timerTask = Task {
                while !Task.isCancelled && viewModel.sosState.step == .timer {
                    try? await Task.sleep(nanoseconds: 1_000_000_000)
                    await MainActor.run { viewModel.tickTimer() }
                }
            }
        }
        if viewModel.sosState.step == .breathing {
            breathingTask = Task {
                while !Task.isCancelled && viewModel.sosState.step == .breathing {
                    try? await Task.sleep(nanoseconds: 4_000_000_000)
                    await MainActor.run { viewModel.tickBreathing() }
                }
            }
        }
    }

    private func stopTimers() {
        timerTask?.cancel()
        breathingTask?.cancel()
    }
}

private struct EmergencyScreen: View {
    let contact: SosContact?
    let onCall: () -> Void
    let onSms: () -> Void
    let onWhatsApp: () -> Void
    let onTelegram: () -> Void
    let onBreathing: () -> Void
    let onTimer: () -> Void
    let onBack: () -> Void

    var body: some View {
        SosFullscreenBackground {
            ScrollView {
                VStack(spacing: 12) {
                    Text("Критическая тяга")
                        .font(.title.bold())
                        .foregroundStyle(.white)
                    if let contact {
                        Text(contact.name).foregroundStyle(.white.opacity(0.8))
                    }
                    emergencyButton("Позвонить", action: onCall)
                    emergencyButton("SMS", action: onSms)
                    emergencyButton("WhatsApp", action: onWhatsApp)
                    emergencyButton("Telegram", action: onTelegram)
                    emergencyButton("Дыхание", action: onBreathing)
                    emergencyButton("⏱ 3 минуты вместе", action: onTimer)
                    Button("Назад", action: onBack)
                        .foregroundStyle(.white.opacity(0.8))
                        .padding(.top, 16)
                }
                .padding(24)
            }
        }
    }

    private func emergencyButton(_ title: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(title)
                .font(.body.weight(.semibold))
                .foregroundStyle(.white)
                .frame(maxWidth: .infinity)
                .padding()
                .background(Color.white.opacity(0.15))
                .clipShape(RoundedRectangle(cornerRadius: 12))
        }
        .buttonStyle(.plain)
    }
}
